polars/polars-lazy/polars-plan/src/dsl/function_expr/schema.rs

```diff
@@ -286,16 +172,132 @@ impl FunctionExpr {
                 })
             }
             #[cfg(feature = "dot_product")]
-            Dot => map_dtype(&|dt| {
+            Dot => mapper.map_dtype(|dt| {
                 use DataType::*;
                 match dt {
                     Int8 | Int16 | UInt16 | UInt8 => Int64,
                     _ => dt.clone(),
                 }
             }),
             #[cfg(feature = "log")]
-            Entropy { .. } | Log { .. } | Log1p | Exp => float_dtype(),
-            Unique(_) => same_type(),
+            Entropy { .. } | Log { .. } | Log1p | Exp => mapper.map_to_float_dtype(),
+            Unique(_) => mapper.with_same_dtype(),
+        }
+    }
+}
+
+pub(super) struct FieldsMapper<'a> {
+    fields: &'a [Field],
+}
+
+impl<'a> FieldsMapper<'a> {
+    /// Field with the same dtype.
+    pub(super) fn with_same_dtype(&self) -> PolarsResult<Field> {
+        self.map_dtype(|dtype| dtype.clone())
+    }
+
+    /// Set a dtype.
+    pub(super) fn with_dtype(&self, dtype: DataType) -> PolarsResult<Field> {
+        Ok(Field::new(self.fields[0].name(), dtype))
+    }
+
+    /// Map a single dtype.
+    pub(super) fn map_dtype(&self, func: impl Fn(&DataType) -> DataType) -> PolarsResult<Field> {
+        let dtype = func(self.fields[0].data_type());
+        Ok(Field::new(self.fields[0].name(), dtype))
+    }
+
+    /// Map to a float supertype.
+    pub(super) fn map_to_float_dtype(&self) -> PolarsResult<Field> {
+        self.map_dtype(|dtype| match dtype {
+            DataType::Float32 => DataType::Float32,
+            _ => DataType::Float64,
+        })
+    }
+
+    /// Map a single dtype with a potentially failing mapper function.
+    #[cfg(feature = "timezones")]
+    pub(super) fn try_map_dtype(
+        &self,
+        func: impl Fn(&DataType) -> PolarsResult<DataType>,
+    ) -> PolarsResult<Field> {
+        let dtype = func(self.fields[0].data_type())?;
+        Ok(Field::new(self.fields[0].name(), dtype))
+    }
+
+    /// Map all dtypes with a potentially failing mapper function.
+    pub(super) fn try_map_dtypes(
+        &self,
+        func: impl Fn(&[&DataType]) -> PolarsResult<DataType>,
+    ) -> PolarsResult<Field> {
+        let mut fld = self.fields[0].clone();
+        let dtypes = self
+            .fields
+            .iter()
+            .map(|fld| fld.data_type())
+            .collect::<Vec<_>>();
+        let new_type = func(&dtypes)?;
+        fld.coerce(new_type);
+        Ok(fld)
+    }
+
+    /// Map the dtype to the "supertype" of all fields.
+    pub(super) fn map_to_super_dtype(&self) -> PolarsResult<Field> {
```

Comment:
Can you name it supertype ? That's consistent with what we do on other locations.